Check core reactions for collision rate limit violators

alongd opened the pull request:
This collision rate limit check is inspired by [Wang et at. 2017](https://doi.org/10.1016/j.combustflame.2017.08.005).
Currently we only output a warning along with a relevant log file if any core reactions violate this limit.

alongd commented:
@amarkpayne, could you take a look at this PR?

alongd commented:
@amarkpayne, I just rebased (conflicts were solved)

amarkpayne commented:
Thanks! I am looking over the code now

amarkpayne commented on rmgpy/rmg/main.py:
```diff
@@ -898,6 +902,27 @@ def execute(self, **kwargs):
+        for rxn in self.reactionModel.core.reactions:
+            f_violator, r_violator = rxn.check_collision_limit_violation(t_min=self.Tmin, t_max=self.Tmax,
+                                                            p_min=self.Pmin, p_max=self.Pmax)
+            if f_violator is not None:
```
I think this statement can be simplified to `if f_violator:`

cgrambow replied:
`if x` and `if x is not None` are used in different contexts. Check the PEP 8 recommendations [here](https://legacy.python.org/dev/peps/pep-0008/#programming-recommendations).

alongd replied:
Thanks @cgrambow, it was helpful reading this piece. I'll leave the `is not None` statement as as

amarkpayne commented on rmgpy/rmg/main.py:
```diff
@@ -898,6 +902,27 @@ def execute(self, **kwargs):
+                violators.append(f_violator)
+            if r_violator is not None:
+                violators.append(r_violator)
+        if len(violators):
```
This can be replaced with `if violators:` since empty lists have evaluate to False. This is not a big deal though

amarkpayne commented on rmgpy/rmg/main.py:
```diff
@@ -898,6 +902,27 @@ def execute(self, **kwargs):
+            logging.info("\n")
+            logging.warning("{0} CORE reactions violate the collision rate limit!"
+                            " See the 'collision_rate_violators.log' for details.\n\n".format(len(violators)))
+            with open('collision_rate_violators.log', 'w') as violators_f:
```
Do we need to check if a file with this name already exists? Otherwise, this file will get overwritten if the user restarts the job (that might be okay for this file, I am just checking because I know that we do this for the RMG log file)

alongd replied:
I think that since the model itself (chemkin/cantera formats) is being overwritten, and this output directly relates to the generated model, it's ok the overwrite the collision violators log too. A related question, though, is what if the user re-runs in the same folder, and only the first run generates the collision violators log file... I guess I should delete it if it exists and no violators were found?

amarkpayne commented on rmgpy/reaction.py:
```diff
@@ -1144,6 +1144,121 @@ def ensure_species(self, reactant_resonance=False, product_resonance=True):
+            for condition in conditions:
+                try:
+                    collision_limit.append(self.calculate_coll_limit(temp=condition[0], reverse=False))
+                except ValueError:
```
Under what circumstances does this try statement throw a ValueError? Is it something that is okay to ignore?

alongd replied:
When I tested this branch by inventing ridiculous rates, I got a `ValueError` from `calculate_coll_limit()`. It was a while ago, so I don't remember exactly what calculation caused it to crash (perhaps either `reduced_coll_integral` or `k_coll`?).
Since we decided that at the fist stage we only want to report rate violators, but not crash RMG, I decided to capture errors here and `pass` on them. This is just to be on the safe side and not crushing a job near its end for a reason we're not certain of.

amarkpayne commented on rmgpy/reaction.py:
```diff
@@ -1144,6 +1144,121 @@ def ensure_species(self, reactant_resonance=False, product_resonance=True):
+                trans = spc.getTransportData()
+                sigmas.append(trans.sigma.value_si)
+                epsilons.append(trans.epsilon.value_si)
+                num_of_spcs = len(self.products)
```
Move this out of the for loop. Otherwise, `num_of_spcs` is set to the same value over and over again

alongd replied:
done, thanks!

amarkpayne commented on rmgpy/reaction.py:
```diff
@@ -1144,6 +1144,121 @@ def ensure_species(self, reactant_resonance=False, product_resonance=True):
+        if any([x == 0 for x in sigmas + epsilons]):
+            raise ValueError
+        mean_sigmas = sum(sigmas) / num_of_spcs
+        mean_epsilons = reduce((lambda x, y: x * y), epsilons) ** (1 / sum(epsilons))
```
For the geometric mean for the well depths, should the product be raised to `(1/len(epsilons))` ? I thought the geometric mean is raised to the inverse power of the number of elements, not their sum.

alongd replied:
You're right, thanks for catching my mistake!

alongd commented:
@amarkpayne, thanks for your suggestions!
I addressed the comments in fixup commits.

I also relocated the collision rate violation checks to the `check_model` method.
Last, I slightly changed the order of imports in rmg/main.py according to [PEP 8 recommendations](https://www.python.org/dev/peps/pep-0008/?#imports). Slight modifications to `os` and `np` imports as well.

Let me know if you have additional comments, or whether we're good to squash.

alongd commented:
@amarkpayne, I simplified the code a bit, and added more relevant info to the collision violation log file

amarkpayne commented on rmgpy/reaction.py:
```diff
@@ -1144,6 +1144,99 @@ def ensure_species(self, reactant_resonance=False, product_resonance=True):
+        logging.debug("Checking whether reaction {0} violates the collision rate limit...".format(self))
+        violator = None
+        kf_list = []
+        kr_list = []
```
With the newest simplification, I don't thing kr_list is being used (are you missing the code that handles checking the reverse reaction? I like the direction of this code simplification, but I think it is only checking the forward direction now.

alongd replied:
Can you take a look at line 1180 in this file? I think `kr_list` is used as expected

https://github.com/ReactionMechanismGenerator/RMG-Py/pull/1361/files/fe94464ff3ad69d8b4c12e9950cc39e1dafede75#diff-eb892c0ffd531a3aa16ba76563c6ccc6R1180

amarkpayne replied:
Does anything append on to `kr_list` before line 1180 though? Maybe I am missing something

alongd replied:
You're right. I'll fix it.

alongd commented:
I guess we shouldn't merge this until the 2.2.0 release is out, right?

amarkpayne commented:
I think that was the plan for this PR, but I'll double check

mliu49 commented on rmgpy/rmg/main.py:
```diff
@@ -945,6 +948,34 @@ def check_model(self):
+                                      'Violation condition: {4}\n\n'.format(
+                        violator[0], violator[1], violator[2], violator[3], violator[4]))
+        else:
+            if os.path.isfile('collision_rate_violators.log'):
```
I think you could replicate the behavior of RMG.log and rename the existing log regardless of whether or not violators were found, i.e. move this check before `if violators:`. It might also be better to put OLD at the end of the filename so that it appears right next to the new file in the file list.

mliu49 commented:
To confirm my understanding, this check is run once at the end of the RMG run, and only prints warnings and saves the log file of violations?

I think it could be included in v2.2 if it's ready.

alongd commented:
@mliu49, your understanding is correct.
From testing this out a bit, I expect we would start seeing many collision rate violators...
A simple ethane oxidation run at 1500-2000 K gave ~20 violators. Perhaps we shouldn't include it in the current binaries release, and just let the developers work with it and be aware of the output to improve the database towards the next release?

amarkpayne commented:
This PR will be merged after the next version release

alongd commented:
@amarkpayne, this branch is rebased and ready to merge once all tests pass
